src/coreComponents/denseLinearAlgebra/denseLASolvers.hpp

```diff
@@ -0,0 +1,291 @@
+ */
+template< std::ptrdiff_t N,
+          typename MATRIX_TYPE,
+          typename RHS_TYPE,
+          typename SOL_TYPE >
+GEOS_HOST_DEVICE
+inline
+void solveGaussianElimination( MATRIX_TYPE & A, RHS_TYPE & b, SOL_TYPE && x )
+{
+  static_assert( N > 0, "N must be greater than 0." );
+  internal::checkSizes< N, N >( matrix );
+  internal::checkSizes< N >( b );
+  internal::checkSizes< N >( x );
+
+
+  // Step 1: Transform  into an upper triangular matrix
+
+  // 1.a. Find the pivot
+  for( std::ptrdiff_t i = 0; i < N; ++i )
+  {
+    std::ptrdiff_t max_row = i;
+    for( std::ptrdiff_t k = i + 1; k < N; ++k )
+    {
+      if( std::abs( A[k][i] ) > std::abs( A[max_row][i] ))
+      {
+        max_row = k;
+      }
+    }
+
+    // 1.b. Swap rows
+    for( std::ptrdiff_t k = i; k < N; ++k )
+    {
+      std::swap( A[i][k], A[max_row][k] );
+    }
+    std::swap( b[i], b[max_row] );
+
+    GEOS_ERROR_IF_LT_MSG( LvArray::math::abs( A[i][i] ), LvArray::NumericLimits< real64 >::epsilon, "Singular matrix." );
+
+    // 1.c Eliminate entries below the pivot
+    for( std::ptrdiff_t k = i + 1; k < N; ++k )
+    {
+      real64 const scaling = A[k][i] / A[i][i];
+      for( std::ptrdiff_t j = i; j < N; ++j )
+      {
+        A[k][j] -= scaling * A[i][j];
+      }
+      b[k] -= scaling * b[i];
+    }
+  }
+
+  // Step 2: Backward substitution
+  solveUpperTriangularSystem< N >( A, b, std::forward< N >( x ) )
+}
+
+}; // internal namespace
+
+/**
+ * @brief Solves a linear system using the most appropriate method based on the size of the system.
+ *
+ * This function determines the appropriate method for solving a linear system `Ax = b` based on
+ * the size of the matrix `A`. For 2x2 and 3x3 systems, specialized solvers are used. For larger systems,
+ * Gaussian elimination is employed. The matrix and the rhs are modified by the function.
+ *
+ * @tparam N The size of the square matrix `A`.
+ * @tparam MATRIX_TYPE The type of the matrix `A`.
+ * @tparam RHS_TYPE The type of the right-hand side vector `b`.
+ * @tparam SOL_TYPE The type of the solution vector `x`.
+ * @tparam MODIFY_MATRIX Boolean flag indicating whether the input matrix `A` and vector `b` should be modified.
+ *                       If `true`, the matrix `A` and vector `b` are modified in place. If `false`, copies of 
+ *                       `A` and `b` are made, and the original data is left unchanged.
+ * @param[in] A The constant matrix representing the coefficients of the system.
+ * @param[in] b The constant right-hand side vector.
+ * @param[out] x The solution vector. The result of solving the system `Ax = b`.
+ */
+template< std::ptrdiff_t N,
+          typename MATRIX_TYPE,
+          typename RHS_TYPE,
+          typename SOL_TYPE,
+          bool MODIFY_MATRIX = true >
+GEOS_HOST_DEVICE
+inline
+void solve( MATRIX_TYPE & A, RHS_TYPE & b, SOL_TYPE && x )
+{
+  static_assert( N > 0, "N must be greater than 0." );
+  static_assert( N < 10, "N cannot be larger than 9" );
```

Comment:
For the PVT application, we may have a size of 2*NC+1 where NC is number of components. I think locally (with PVT driver) we allow up to 9 components.

Reply:
Just seen the comments above about registers. Maybe I can "manually" reduce the matrix size to NC+1.